src/lib/Transpose.hs

```diff
@@ -70,12 +70,21 @@ isLin e = do
 withAccumulator
   :: Emits o
   => Type o
-  -> (forall o'. (Emits o', DExt o o') => Atom o' -> TransposeM i o' ())
+  -> (forall o'. (Emits o', DExt o o') => TransposeSubstVal AtomNameC o' -> TransposeM i o' ())
   -> TransposeM i o (Atom o)
 withAccumulator ty cont = do
-  baseMonoid <- getBaseMonoidType ty >>= tangentBaseMonoidFor
-  getSnd =<< emitRunWriter noHint ty baseMonoid \_ ref ->
-               cont (Var ref) >> return UnitVal
+  singletonTypeVal ty >>= \case
+    Nothing -> do
+      baseMonoid <- getBaseMonoidType ty >>= tangentBaseMonoidFor
```

Comment:
nit: it would read more like English if you did `tangentBaseMonoidFor =<< getBaseMonoidType ty`